### PR TITLE
Introduce new product form structure for BO Experimental Product page

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Change the structure of the product form, the basic tab has been split into two tabs and some fields from options have been moved into the new Specifications tab<br>Also adapt the data provider and the commands builder and their matching tests<br>The new header now includes the cover image and some resume of the references<br>The command builder for product references was never implemented so it's been added into this PR as well<br>The shortcut column in the first tab was removed, for now only the form field and twig part have been removed, the form types and js code are still present for now We want for the validation of the new page layout to remove them (just in case)
| Type?             | improvement
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | ~
| How to test?      | The experimental product page should (mostly) match the new structure from the mockups here https://miro.com/app/board/o9J_l7kJKl4=/ Some parts are not implemented yet, or will be refactored in another PR (like categories, prices in header, automatic update of header, ...)
| Possible impacts? | ~

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25419)

## Breaking Changes

- `ProductForEditing` now contains a `coverUrl` field, `GetProductForEditingHandler` was also enriched with new services injected to get this new data
- Form structure has been modified so the data is not split the same way now, additionally some form types (and therefore their services) have been renamed to match the new structure (like `BasicType` is now `DescriptionType`) or were moved in a new namespace

### CODE REVIEW

File: src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php [hunk 1]
```diff
@@ -391,4 +411,14 @@ private function getVirtualProductFile(Product $product): ?VirtualProductFileFor
             $virtualProductFile->date_expiration === DateTimeUtil::NULL_DATETIME ? null : new DateTime($virtualProductFile->date_expiration)
         );
     }
+
+    private function getCover(Product $product): string
```
Comment (kpodemski) [inline src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php:420]: Missing docs

File: src/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilder.php [hunk 2]
```diff
@@ -41,21 +41,21 @@ class BasicInformationCommandsBuilder implements ProductCommandsBuilderInterface
...
-        $basicData = $formData['basic'] ?? [];
+        $descriptionData = $formData['description'] ?? [];
         $command = new UpdateProductBasicInformationCommand($productId->getValue());
 
         if (isset($formData['header']['name'])) {
             $command->setLocalizedNames($formData['header']['name']);
         }
-        if (isset($basicData['description'])) {
-            $command->setLocalizedDescriptions($basicData['description']);
+        if (isset($descriptionData['description'])) {
```
Comment (kpodemski) [inline src/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilder.php:54]: So the original path is `$formData['description']['description']`?

Comment (kpodemski) [reply src/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilder.php:54]: If so, I'd rather use something like `$formData['informations']`?

File: src/Core/Form/IdentifiableObject/CommandBuilder/Product/DetailsCommandsBuilder.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Form\IdentifiableObject\CommandBuilder\Product;
+
+use PrestaShop\PrestaShop\Core\Domain\Product\Command\UpdateProductDetailsCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+
+class DetailsCommandsBuilder implements ProductCommandsBuilderInterface
+{
+    public function buildCommands(ProductId $productId, array $formData): array
```
Comment (kpodemski) [inline src/Core/Form/IdentifiableObject/CommandBuilder/Product/DetailsCommandsBuilder.php:42]: Missing docblock

File: tests/Integration/Behaviour/Features/Scenario/Product/Image/update_image.feature [hunk 4]
```diff
@@ -104,6 +108,7 @@ Feature: Update product image from Back Office (BO)
       | image4          | false    |               | 4        | http://myshop.com/img/p/{image4}.jpg | http://myshop.com/img/p/{image4}-small_default.jpg |
       | image5          | false    |               | 5        | http://myshop.com/img/p/{image5}.jpg | http://myshop.com/img/p/{image5}-small_default.jpg |
       | image2          | false    |               | 6        | http://myshop.com/img/p/{image2}.jpg | http://myshop.com/img/p/{image2}-small_default.jpg |
+    And product "product1" should have following cover "http://myshop.com/img/p/{image1}.jpg"
```
Comment (kpodemski) [inline tests/Integration/Behaviour/Features/Scenario/Product/Image/update_image.feature]: I didn't notice this before, in this tab, do we load an original image? maybe it would be better to load some type of image? to decrease load time

File: tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilderTest.php [hunk 5]
```diff
@@ -58,7 +58,7 @@ public function getExpectedCommands()
         $command = new UpdateProductBasicInformationCommand($this->getProductId()->getValue());
         yield [
             [
-                'basic' => [
+                'description' => [
```
Comment (kpodemski) [inline tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilderTest.php:61]: Even command name is `BasicInformation` tho you decide to use `description`? :))

File: tests/Integration/Behaviour/Features/Scenario/Product/Image/update_image.feature (hunk 4 above)
Comment (Hlavtox) [reply tests/Integration/Behaviour/Features/Scenario/Product/Image/update_image.feature]: @kpodemski Well, that's the only one we know for sure that it exists... Some thumbnail would be better, but you need to make sure:
A) Its generated. (Not sure if it is fixed already, on-the-fly generation didn't work for PNGs)
B) This thumbnail size is registered and used for products. 

Try to think with the team if it wouldn't be good to make some image sizes undeletable, like some order statuses are. At least the ones used in BO should be like this.

File: tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/DetailsCommandsBuilderTest.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+class DetailsCommandsBuilderTest extends AbstractProductCommandBuilderTest
+{
+    /**
+     * @dataProvider getExpectedCommands
+     *
+     * @param array $formData
+     * @param array $expectedCommands
+     */
+    public function testBuildCommand(array $formData, array $expectedCommands)
```
Comment (Progi1984) [inline tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/DetailsCommandsBuilderTest.php]:
```suggestion
    public function testBuildCommand(array $formData, array $expectedCommands): void
```

File: tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/DetailsCommandsBuilderTest.php [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    {
+        $builder = new DetailsCommandsBuilder();
+        $builtCommands = $builder->buildCommands($this->getProductId(), $formData);
+        $this->assertEquals($expectedCommands, $builtCommands);
+    }
+
+    public function getExpectedCommands()
```
Comment (Progi1984) [inline tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/DetailsCommandsBuilderTest.php]:
```suggestion
    public function getExpectedCommands(): iterable
```

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml [hunk 8]
```diff
@@ -1528,20 +1549,6 @@ services:
         tags:
             - { name: form.type }
 
-    form.type.sell.product.options.references_type:
```
Comment (Progi1984) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:1531]: BC Break : services existing in 178x

File: src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php [hunk 9]
```diff
@@ -124,7 +125,7 @@ public function getDefaultData()
             'header' => [
                 'type' => ProductType::TYPE_STANDARD,
             ],
-            'basic' => [
+            'description' => [
```
Comment (Progi1984) [inline src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php:127]: What are default data for `specifications` ?

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml (hunk 8 above)
Comment (kpodemski) [reply src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:1531]: if it's introduced in non released version, and then removed, why it's BC Break?

Comment (kpodemski) [reply src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:1531]: nevermind, it will be available in 1.7.8, and then removed in 8.0.0 👍

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:1531]: Besides it's just a form type definition, form types are never accessed directly via the container And this one is part of the experimental page so we are free to break/modify anything we want

But I'll add this into the long list of things I already broke if it's important for you ^^

File: src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php (hunk 9 above)
Comment (jolelievre) [reply src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php:127]: No default data, they are empty like most of the data

File: src/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilder.php (hunk 2 above)
Comment (jolelievre) [reply src/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilder.php:54]: Yes, I would also prefer to use `informations` to avoid having it twice, but the main reason here is that the name of the first form types, match the names of the tabs I think it's the easiest way to keep a logical meaning and to find the appropriate path more easily

In the new page structure the former tab `Basic settings` has been split into two tabs `Description` and `Specifications` (see screenshot)

Capture d’écran 2021-07-26 à 11 01 11

So keeping this logic seems more important than avoiding duplicity. Regarding the CQRS commands they are less coupled with presentation than the form so that's why I didn't rename them for now, but that's a possibility I don't completely ignore. Maybe some commands will be renamed to keep the consistency, we didn't decide yet.

So TLDR; if the tab was named `Informations` I wouldn't mind using the same wording in the form type, but that's a decision from @PrestaShop/product-team @MatShir or @TristanLDD

File: tests/Integration/Behaviour/Features/Scenario/Product/Image/update_image.feature (hunk 4 above)
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Product/Image/update_image.feature]: Theoretically, all thumbnails are indeed present and available, as long as the image type has not been removed. We are indeed thinking of preventing the removal of basic image types, but that's another topic.

In this case the image is used in the new Product form where we display the cover in the header to help the merchant identify immediately which product he is editing:

Capture d’écran 2021-07-26 à 11 09 19

The graphical display is not definitive as for now we only have some functional mock-ups available, but that's the generic idea.

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Product/Image/update_image.feature]: So I don't know which image type I should use, I thought that by default I could use the original one to be sure the quality is good, but maybe it would be better to use a thumbnail. Since it's displayed in the BO I'm not sure the download cost is that heavy, but we can always modify this later anyway.

That said if you think it's preferable to get the thumbnail right now I can do the modif, using the `small_default` type?

File: tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilderTest.php (hunk 5 above)
Comment (jolelievre) [reply tests/Unit/Core/Form/IdentifiableObject/CommandBuilder/Product/BasicInformationCommandsBuilderTest.php:61]: `BasicInformation` was mostly named after the tab name back then, as I explained in another comment the tab is now called description, hence the renaming of the form field, and maybe the command should also be renamed.

File: admin-dev/themes/new-theme/js/pages/product/edit/product-form-mapping.js [hunk 10]
```diff
@@ -26,19 +26,19 @@
 export default {
   'product.stock.quantity': [
     'product[stock][quantities][quantity]',
-    'product[shortcuts][stock][quantity]',
+    // 'product[shortcuts][stock][quantity]',
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/product/edit/product-form-mapping.js:29]: Why do you comment this lines ?

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-form-mapping.js:29]: Because there is no shortcut column anymore in the new UX, as I explained in the description of this PR for now it is more "disabled" than completely removed, so the code has not been removed yet

This JS code caused bugs in the console because the shortcuts fields are not present anymore so I just commented this part, it will be removed along with the other code related to shortcut later but it doesn't mean this first modifications can't be merged

Once the page is more mature and Tristan validates the new display I will create a dedicated PR to clean everything correctly

Comment (Progi1984) [reply admin-dev/themes/new-theme/js/pages/product/edit/product-form-mapping.js:29]: :+1:

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — APPROVED** [review]
LGTM

It's not such a big change :) I'm OK with it.
And we can always write a doc about the form structure to summarize all the "where is my data ? where is the input for field XYZ ?"

**Review by Progi1984 — APPROVED** [review]

**Comment by kpodemski** [comment]
@PrestaShop/prestashop-maintainers is QA really needed now? I think it has to be merged for others to continue their work and prepare full QA at some point?

**Comment by matks** [comment]
> @PrestaShop/prestashop-maintainers is QA really needed now? I think it has to be merged for others to continue their work and prepare full QA at some point?

Indeed, with @Robin-Fischer-PS we agreed that on new Product page, QA will test complete features and not single PRs. @jolelievre is this a complete feature, or will there be more PRs for this topic?

**Comment by matks** [comment]
Note to myself: @Robin-Fischer-PS is enjoying some well-merited vacations 😄  .

Seen with @jolelievre, we think this can be merged without QA as agreed with @Robin-Fischer-PS : other PRs are to come to implement missing parts and refactor other parts.

When this is complete, QA will be able to validate the full deliverable, and not this PR that only delivers a partial feature.
